src/microbe_stage/ProcessSystem.cs

```diff
@@ -377,81 +381,115 @@ private void ProcessNode(IProcessable processor, float delta, float inverseDelta
 
-            var inputRemoved = entry.Value * process.Rate * delta;
-
-            currentProcessStatistics?.AddInputAmount(entry.Key, inputRemoved * inverseDelta);
+            var inputRemoved = entry.Value * process.Rate * delta * environmentModifier * spaceConstraintModifier;
 
             // currentProcessStatistics?.AddInputAmount(entry.Key, 0);
+            // We don't multiply by delta here because we report the per-second values anyway. In the actual process
+            // output numbers (computed after testing the speed), we need to multiply by inverse delta
+            currentProcessStatistics?.AddInputAmount(entry.Key, entry.Value * process.Rate * environmentModifier);
 
-            // If not enough compound we can't do the process
-            if (bag.GetCompoundAmount(entry.Key) < inputRemoved)
+            // If not enough we can't run the process unless we can lower spaceConstraintModifier enough
+            var availableAmount = bag.GetCompoundAmount(entry.Key);
+            if (availableAmount < inputRemoved)
             {
-                canDoProcess = false;
-                currentProcessStatistics?.AddLimitingFactor(entry.Key);
+                bool canRun = false;
+
+                if (availableAmount > MathUtils.EPSILON)
+                {
+                    var neededModifier = availableAmount / inputRemoved;
+
+                    if (neededModifier > Constants.MINIMUM_RUNNABLE_PROCESS_FRACTION)
+                    {
+                        spaceConstraintModifier = neededModifier;
+                        canRun = true;
+
+                        // Due to rounding errors there can be very small disparity here between the amount available
+                        // and what we will take with the modifiers. See the comment in outputs for more details
+                    }
+                }
+
+                if (!canRun)
+                {
+                    canDoProcess = false;
+                    currentProcessStatistics?.AddLimitingFactor(entry.Key);
+                }
             }
         }
 
-        // Output
-        // This is now always looped (even when we can't do the process)
-        // because the is useful part is needs to be always be done
         foreach (var entry in processData.Outputs)
         {
-            // For now lets assume compounds we produce are also
-            // useful
+            // For now lets assume compounds we produce are also useful
             bag.SetUseful(entry.Key);
 
-            // Apply the general modifiers and
-            // apply the environmental modifier
-            var outputAdded = entry.Value * process.Rate * delta * environmentModifier;
-
-            currentProcessStatistics?.AddOutputAmount(entry.Key, outputAdded * inverseDelta);
+            var outputAdded = entry.Value * process.Rate * delta * environmentModifier * spaceConstraintModifier;
 
             // currentProcessStatistics?.AddOutputAmount(entry.Key, 0);
```

Comment:
```suggestion
```

also looks like old code that can go now.

Reply:
https://github.com/Revolutionary-Games/Thrive/pull/2488#discussion_r688922023